usr.bin/procstat/procstat.1
.\"-
.\" Copyright (c) 2007-2009 Robert N. M. Watson
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD$
.\"
<<<<<<< HEAD
.Dd October 14, 2017
=======
.Dd October 3, 2017
>>>>>>> 930409367ecf72a59ee5666730e1b84ac90527b2
.Dt PROCSTAT 1
.Os
.Sh NAME
.Nm procstat
.Nd get detailed process information
.Sh SYNOPSIS
.Nm
.Op Fl -libxo
<<<<<<< HEAD
.Op Fl h
.Op Fl M Ar core
.Op Fl N Ar system
.Op Fl w Ar interval
.Ar command
.Op Ar pid ... | Ar core ...
.Nm
.Op Fl -libxo
.Fl a
.Op Fl h
.Op Fl M Ar core
.Op Fl N Ar system
.Op Fl w Ar interval
.Ar command
.Nm
.Op Fl -libxo
.Op Fl h
.Op Fl M Ar core
.Op Fl N Ar system
.Op Fl w Ar interval
.Oo
.Fl b |
.Fl c |
.Fl e |
.Fl f Oo Fl C Oc |
.Fl i Oo Fl n Oc |
.Fl j Oo Fl n Oc |
.Fl k Oo Fl k Oc |
.Fl l |
.Fl r Oo Fl H Oc |
.Fl s |
.Fl S |
.Fl t |
.Fl v |
.Fl x
.Oc
.Op Ar pid ... | Ar core ...
.Nm
.Op Fl -libxo
.Fl a
.Op Fl h
.Op Fl M Ar core
.Op Fl N Ar system
.Op Fl w Ar interval
.Oo
.Fl b |
.Fl c |
.Fl e |
.Fl f Oo Fl C Oc |
.Fl i Oo Fl n Oc |
.Fl j Oo Fl n Oc |
.Fl k Oo Fl k Oc |
.Fl l |
.Fl r Oo Fl H Oc |
.Fl s |
.Fl S |
.Fl t |
.Fl v |
.Fl x
.Oc
.Nm
.Op Fl -libxo
.Fl L
.Op Fl h
.Op Fl M Ar core
.Op Fl N Ar system
.Op Fl w Ar interval
.Ar core ...
=======
.Op Fl CHhn
.Op Fl M Ar core
.Op Fl N Ar system
.Op Fl w Ar interval
.Op Fl b | c | e | f | i | j | k | l | L | r | s | S | t | v | x
.Op Fl a | Ar pid | Ar core ...
>>>>>>> 930409367ecf72a59ee5666730e1b84ac90527b2
.Sh DESCRIPTION
.Nm
utility displays detailed information about the processes identified by the
.Ar pid
arguments, or if the
.Fl a
flag is used, all processes.
It can also display information extracted from a process core file, if
the core file is specified as the argument.
.Pp
If the
.Fl -libxo
flag is specified the output is generated via
.Xr libxo 3
in a selection of different human and machine readable formats.
See
.Xr xo_parse_args 3
for details on command line arguments.
.Pp
The following commands are available:
.Bl -tag -width indent
.It Ar basic
Print basic process statistics (this is the default).
.It Ar binary | Fl b
Display binary information for the process.
.Pp
Substring commands are accepted.
.It Ar argument(s) | Fl c
Display command line arguments for the process.
.Pp
Substring commands are accepted.
.It Ar environment | Fl e
Display environment variables for the process.
.Pp
Substring commands are accepted.
.It Ar file(s) | Ar fd(s) | Fl f
Display file descriptor information for the process.
.Pp
If the
.Fl C
subcommand flag is used then additional capability information is printed.
.It Ar signal(s) | Fl i
Display signal pending and disposition information for the process.
.Pp
If the
.Fl n
subcommand option is used, the signal numbers are shown instead of signal
names.
.Pp
Substring commands are accepted.
.It Ar tsignal(s) | Fl j
Display signal pending and blocked information for the process's threads.
.Pp
If the
.Fl n
subcommand option is used, the signal numbers are shown instead of signal
names.
.Pp
Substring commands are accepted.
.It Ar kstack | Fl k
Display the stacks of kernel threads in the process, excluding stacks of
threads currently running on a CPU and threads with stacks swapped to disk.
.Pp
If the
.Fl v
subcommand option is used (or the command flag is repeated), function
offsets as well as function names are printed.
.It Ar rlimit | Fl l
Display resource limits for the process.
<<<<<<< HEAD
.It Ar ptlwpinfo | Fl L
Display LWP info for the process pertaining to its signal driven exit.
.It Ar rusage | Fl r
=======
.It Fl L
Display LWP info for the process pertaining to it's signal driven exit.
.It Fl r
>>>>>>> 930409367ecf72a59ee5666730e1b84ac90527b2
Display resource usage information for the process.
.Pp
If the
.Fl v
.Pq or Fl H
subcommand flag
is used then per-thread statistics are printed, rather than per-process
statistics.
The second field in the table will list the thread ID to which the row of
information corresponds.
.It Ar credential(s) | Fl s
Display security credential information for the process.
.Pp
Substring commands are accepted.
.It Ar cpuset | Ar cs | Fl S
Display the cpuset information for the thread.
.It Ar thread(s) | Fl t
Display thread information for the process.
.It Ar vm | Fl v
Display virtual memory mappings for the process.
.It Ar auxv | Fl x
Display ELF auxiliary vector for the process.
.El
.Pp
All options generate output in the format of a table, the first field of
which is the process ID to which the row of information corresponds.
The
.Fl h
flag may be used to suppress table headers.
.Pp
The
.Fl w
flag may be used to specify a wait interval at which to repeat the printing
of the requested process information.
If the
.Fl w
flag is not specified, the output will not repeat.
.Pp
Information for VM, file descriptor, and cpuset options is available
only to the owner of a process or the superuser.
A cpuset value displayed as -1 means that the information is either invalid
or not available.
.Ss Binary Information
Display the process ID, command, and path to the process binary:
.Pp
.Bl -tag -width indent -compact
.It PID
process ID
.It COMM
command
.It OSREL
osreldate for process binary
.It PATH
path to process binary (if available)
.El
.Ss Command Line Arguments
Display the process ID, command, and command line arguments:
.Pp
.Bl -tag -width indent -compact
.It PID
process ID
.It COMM
command
.It ARGS
command line arguments (if available)
.El
.Ss Environment Variables
Display the process ID, command, and environment variables:
.Pp
.Bl -tag -width "ENVIRONMENT" -compact
.It PID
process ID
.It COMM
command
.It ENVIRONMENT
environment variables (if available)
.El
.Ss File Descriptors
Display detailed information about each file descriptor referenced by a
process, including the process ID, command, file descriptor number, and
per-file descriptor object information, such as object type and file system
path.
By default, the following information will be printed:
.Pp
.Bl -tag -width indent -compact
.It PID
process ID
.It COMM
command
.It FD
file descriptor number or cwd/root/jail
.It T
file descriptor type
.It V
vnode type
.It FLAGS
file descriptor flags
.It REF
file descriptor reference count
.It OFFSET
file descriptor offset
.It PRO
network protocol
.It NAME
file path or socket addresses (if available)
.El
.Pp
The following file descriptor types may be displayed:
.Pp
.Bl -tag -width X -compact
.It c
crypto
.It e
POSIX semaphore
.It f
fifo
.It h
shared memory
.It k
kqueue
.It m
message queue
.It P
process descriptor
.It p
pipe
.It s
socket
.It t
pseudo-terminal master
.It v
vnode
.El
.Pp
The following vnode types may be displayed:
.Pp
.Bl -tag -width X -compact
.It -
not a vnode
.It b
block device
.It c
character device
.It d
directory
.It f
fifo
.It l
symbolic link
.It r
regular file
.It s
socket
.It x
revoked device
.El
.Pp
The following file descriptor flags may be displayed:
.Pp
.Bl -tag -width X -compact
.It r
read
.It w
write
.It a
append
.It s
async
.It f
fsync
.It n
non-blocking
.It d
direct I/O
.It l
lock held
.El
.Pp
If the
.Fl C
flag is specified, the vnode type, reference count, and offset fields will be
omitted, and a new capabilities field will be included listing capabilities,
as described in
.Xr cap_rights_limit 2 ,
present for each capability descriptor.
.Pp
The following network protocols may be displayed (grouped by address family):
.Pp
.Dv AF_INET ,
.Dv AF_INET6
.Pp
.Bl -tag -width indent -compact
.It ICM
.Dv IPPROTO_ICMP ;
see
.Xr icmp 4 .
.It IPD
.Dv IPPROTO_DIVERT ;
see
.Xr divert 4 .
.It IP?
unknown protocol.
.It RAW
.Dv IPPROTO_RAW ;
see
.Xr ip 4 .
.It SCT
.Dv IPPROTO_SCTP ;
see
.Xr sctp 4 .
.It TCP
.Dv IPPROTO_TCP ;
see
.Xr tcp 4 .
.It UDP
.Dv IPPROTO_UDP ;
see
.Xr udp 4 .
.El
.Pp
.Dv AF_LOCAL
.Pp
.Bl -tag -width indent -compact
.It UDD
.Dv IPPROTO_UDP ;
see
.Xr udp 4 .
.It UDS
.Dv IPPROTO_TCP ;
see
.Xr tcp 4 .
.It UD?
unknown protocol.
.El
.Pp
.Bl -tag -width indent -compact
.It ?
unknown address family.
.El
.Ss Signal Disposition Information
Display signal pending and disposition for a process:
.Pp
.Bl -tag -width indent -compact
.It PID
process ID
.It COMM
command
.It SIG
signal name
.It FLAGS
process signal disposition details, three symbols
.Bl -tag -width X -compact
.It P
if signal is pending in the global process queue; - otherwise.
.It I
if signal delivery disposition is
.Dv SIG_IGN;
- otherwise.
.It C
if the signal will be caught; - otherwise.
.El
.El
.Pp
If
.Fl n
switch is given, the signal numbers are shown instead of signal names.
.Ss Thread Signal Information
Display signal pending and blocked for a process's threads:
.Pp
.Bl -tag -width indent -compact
.It PID
process ID
.It TID
thread ID
.It COMM
command
.It SIG
signal name
.It FLAGS
thread signal delivery status, two symbols
.Bl -tag -width X -compact
.It P
if signal is pending for the thread, - otherwise
.It B
if signal is blocked in the thread signal mask, - if not blocked
.El
.El
.Pp
The
.Fl n
switch has the same effect as for the
.Fl i
switch: the signal numbers are shown instead of signal names.
.Ss Kernel Thread Stacks
Display kernel thread stacks for a process, allowing further interpretation
of thread wait channels.
If the
.Fl k
flag is repeated, function offsets, not just function names, are printed.
.Pp
This feature requires
.Cd "options STACK"
or
.Cd "options DDB"
to be compiled into the kernel.
.Pp
.Bl -tag -width indent -compact
.It PID
process ID
.It TID
thread ID
.It COMM
command
.It TDNAME
thread name
.It KSTACK
kernel thread call stack
.El
.Ss Resource Limits
Display resource limits for a process:
.Pp
.Bl -tag -width indent -compact
.It PID
process ID
.It COMM
command
.It RLIMIT
resource limit name
.It SOFT
soft limit
.It HARD
hard limit
.El
.Ss Resource Usage
Display resource usage for a process.
If the
.Fl H
flag is specified,
resource usage for individual threads is displayed instead.
.Pp
.Bl -tag -width "RESOURCE" -compact
.It PID
process ID
.It TID
thread ID
.Po
if
.Fl H
is specified
.Pc
.It COMM
command
.It RESOURCE
resource name
.It VALUE
current usage
.El
.Ss Security Credentials
Display process credential information:
.Pp
.Bl -tag -width indent -compact
.It PID
process ID
.It COMM
command
.It EUID
effective user ID
.It RUID
real user ID
.It SVUID
saved user ID
.It EGID
effective group ID
.It RGID
real group ID
.It SVGID
saved group ID
.It UMASK
file creation mode mask
.It FLAGS
credential flags
.It GROUPS
group set
.El
.Pp
The following credential flags may be displayed:
.Pp
.Bl -tag -width X -compact
.It C
capability mode
.El
.Ss Thread Information
Display per-thread information, including process ID, per-thread ID, name,
CPU, and execution state:
.Pp
.Bl -tag -width indent -compact
.It PID
process ID
.It TID
thread ID
.It COMM
command
.It TDNAME
thread name
.It CPU
current or most recent CPU run on
.It PRI
thread priority
.It STATE
thread state
.It WCHAN
thread wait channel
.El
.Ss Virtual Memory Mappings
Display process virtual memory mappings, including addresses, mapping
meta-data, and mapped object information:
.Pp
.Bl -tag -width indent -compact
.It PID
process ID
.It START
starting address of mapping
.It END
ending address of mapping
.It PRT
protection flags
.It RES
resident pages
.It PRES
private resident pages
.It REF
reference count
.It SHD
shadow page count
.It FLAG
mapping flags
.It TP
VM object type
.El
.Pp
The following protection flags may be displayed:
.Pp

usr.bin/procstat/procstat.c
/*-
 * SPDX-License-Identifier: BSD-2-Clause-FreeBSD
 *
 * Copyright (c) 2007, 2011 Robert N. M. Watson
 * Copyright (c) 2015 Allan Jude <allanjude@freebsd.org>
 * Copyright (c) 2017 Dell EMC
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * $FreeBSD$
 */

#include <sys/param.h>
#include <sys/sysctl.h>
#include <sys/user.h>

#include <err.h>
#include <libprocstat.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sysexits.h>
#include <unistd.h>

#include "procstat.h"

<<<<<<< HEAD
enum {
	PS_CMP_NORMAL = 0x00,
	PS_CMP_PLURAL = 0x01, 
	PS_CMP_SUBSTR = 0x02
};

struct procstat_cmd {
	const char *command;
	const char *xocontainer;
	const char *usage;
	void (*cmd)(struct procstat *, struct kinfo_proc *);
	void (*opt)(int, char * const *);
	int cmp;
};

int procstat_opts = 0;

static void cmdopt_none(int argc, char * const argv[]);
static void cmdopt_verbose(int argc, char * const argv[]);
static void cmdopt_signals(int argc, char * const argv[]);
static void cmdopt_rusage(int argc, char * const argv[]);
static void cmdopt_files(int argc, char * const argv[]);
static void cmdopt_cpuset(int argc, char * const argv[]);

static const struct procstat_cmd cmd_table[] = {
	{ "argument", "arguments", NULL, &procstat_args, &cmdopt_none,
	    PS_CMP_PLURAL | PS_CMP_SUBSTR },
	{ "auxv", "auxv", NULL, &procstat_auxv, &cmdopt_none, PS_CMP_NORMAL },
	{ "basic", "basic", NULL, &procstat_basic, &cmdopt_none,
	    PS_CMP_NORMAL },
	{ "binary", "binary", NULL, &procstat_bin, &cmdopt_none,
	    PS_CMP_SUBSTR },
	{ "cpuset", "cs", NULL, &procstat_cs, &cmdopt_cpuset, PS_CMP_NORMAL },
	{ "cs", "cs", NULL, &procstat_cs, &cmdopt_cpuset, PS_CMP_NORMAL },
	{ "credential", "credentials", NULL, &procstat_cred, &cmdopt_none,
	    PS_CMP_PLURAL | PS_CMP_SUBSTR },
	{ "environment", "environment", NULL, &procstat_env, &cmdopt_none,
	    PS_CMP_SUBSTR },
	{ "fd", "files", "[-C]", &procstat_files, &cmdopt_files,
	    PS_CMP_PLURAL },
	{ "file", "files", "[-C]", &procstat_files, &cmdopt_files,
	    PS_CMP_PLURAL },
	{ "kstack", "kstack", "[-v]", &procstat_kstack, &cmdopt_verbose,
	    PS_CMP_NORMAL },
	{ "ptlwpinfo", "ptlwpinfo", NULL, &procstat_ptlwpinfo, &cmdopt_none,
	    PS_CMP_NORMAL },
	{ "rlimit", "rlimit", NULL, &procstat_rlimit, &cmdopt_none,
	    PS_CMP_NORMAL },
	{ "rusage", "rusage", "[-Ht]", &procstat_rusage, &cmdopt_rusage,
	    PS_CMP_NORMAL },
	{ "signal", "signals", "[-n]", &procstat_sigs, &cmdopt_signals,
	    PS_CMP_PLURAL | PS_CMP_SUBSTR },
	{ "thread", "threads", NULL, &procstat_threads, &cmdopt_none,
	    PS_CMP_PLURAL },
	{ "tsignal", "thread_signals", "[-n]", &procstat_threads_sigs,
	    &cmdopt_signals, PS_CMP_PLURAL | PS_CMP_SUBSTR },
	{ "vm", "vm", NULL, &procstat_vm, &cmdopt_none, PS_CMP_NORMAL }
};
=======
static int aflag, bflag, cflag, eflag, fflag, iflag, jflag, kflag;
static int lflag, Lflag, rflag, sflag, tflag, vflag, xflag, Sflag;
int	hflag, nflag, Cflag, Hflag;
>>>>>>> 930409367ecf72a59ee5666730e1b84ac90527b2

static void
usage(void)
{
<<<<<<< HEAD
	size_t i, l;
	int multi;

	xo_error("usage: procstat [--libxo] [-h] [-M core] [-N system]"
	    " [-w interval] command\n"
	    "                [pid ... | core ...]\n"
	    "       procstat [--libxo] -a [-h] [-M core] [-N system] "
	    " [-w interval] command\n"
	    "       procstat [--libxo] [-h] [-M core] [-N system]"
	    " [-w interval]\n"
	    "                [-S | -b | -c | -e | -f [-C] | -i [-n] | "
	    "-j [-n] | -k [-k] |\n"
	    "                 -l | -r [-H] | -s | -t | -v | -x] "
	    "[pid ... | core ...]\n"
	    "       procstat [--libxo] -a [-h] [-M core] [-N system]"
	    " [-w interval]\n"
	    "                [-S | -b | -c | -e | -f [-C] | -i [-n] | "
	    "-j [-n] | -k [-k] |\n"
	    "                 -l | -r [-H] | -s | -t | -v | -x]\n"
	    "       procstat [--libxo] -L [-h] [-M core] [-N system] core ...\n"
	    "Available commands:\n");
	for (i = 0, l = nitems(cmd_table); i < l; i++) {
		multi = i + 1 < l && cmd_table[i].cmd == cmd_table[i + 1].cmd;
		xo_error("       %s%s%s", multi ? "[" : "",
		    cmd_table[i].command, (cmd_table[i].cmp & PS_CMP_PLURAL) ?
		    "(s)" : "");
		for (; i + 1 < l && cmd_table[i].cmd == cmd_table[i + 1].cmd;
		    i++)
			xo_error(" | %s%s", cmd_table[i + 1].command,
			    (cmd_table[i].cmp & PS_CMP_PLURAL) ? "(s)" : "");
		if (multi)
			xo_error("]");
		if (cmd_table[i].usage != NULL)
			xo_error(" %s", cmd_table[i].usage);
		xo_error("\n");
	}
=======

	xo_error("usage: procstat [--libxo] [-CHhn] [-M core] "
	    "[-N system] [-w interval]\n"
	    "                [-b | -c | -e | -f | -i | -j | -k | "
	    "-l | -r | -s | \n"
	    "                 -S | -t | -v | -x]\n"
	    "                [-a | pid | core ...]\n");
>>>>>>> 930409367ecf72a59ee5666730e1b84ac90527b2
	xo_finish();
	exit(EX_USAGE);
}

static void
procstat(const struct procstat_cmd *cmd, struct procstat *prstat,
    struct kinfo_proc *kipp)
{
	char *pidstr = NULL;

	asprintf(&pidstr, "%d", kipp->ki_pid);
	if (pidstr == NULL)
		xo_errc(1, ENOMEM, "Failed to allocate memory in procstat()");
	xo_open_container(pidstr);
<<<<<<< HEAD
	cmd->cmd(prstat, kipp);
=======

	if (bflag)
		procstat_bin(prstat, kipp);
	else if (cflag)
		procstat_args(prstat, kipp);
	else if (eflag)
		procstat_env(prstat, kipp);
	else if (fflag)
		procstat_files(prstat, kipp);
	else if (iflag)
		procstat_sigs(prstat, kipp);
	else if (jflag)
		procstat_threads_sigs(prstat, kipp);
	else if (kflag)
		procstat_kstack(prstat, kipp, kflag);
	else if (lflag)
		procstat_rlimit(prstat, kipp);
	else if (Lflag)
		procstat_ptlwpinfo(prstat);
	else if (rflag)
		procstat_rusage(prstat, kipp);
	else if (sflag)
		procstat_cred(prstat, kipp);
	else if (tflag)
		procstat_threads(prstat, kipp);
	else if (vflag)
		procstat_vm(prstat, kipp);
	else if (xflag)
		procstat_auxv(prstat, kipp);
	else if (Sflag)
		procstat_cs(prstat, kipp);
	else
		procstat_basic(kipp);

>>>>>>> 930409367ecf72a59ee5666730e1b84ac90527b2
	xo_close_container(pidstr);
	free(pidstr);
}

/*
 * Sort processes first by pid and then tid.
 */
static int
kinfo_proc_compare(const void *a, const void *b)
{
	int i;

	i = ((const struct kinfo_proc *)a)->ki_pid -
	    ((const struct kinfo_proc *)b)->ki_pid;
	if (i != 0)
		return (i);
	i = ((const struct kinfo_proc *)a)->ki_tid -
	    ((const struct kinfo_proc *)b)->ki_tid;
	return (i);
}

void
kinfo_proc_sort(struct kinfo_proc *kipp, int count)
{

	qsort(kipp, count, sizeof(*kipp), kinfo_proc_compare);
}

const char *
kinfo_proc_thread_name(const struct kinfo_proc *kipp)
{
	static char name[MAXCOMLEN+1];

	strlcpy(name, kipp->ki_tdname, sizeof(name));
	strlcat(name, kipp->ki_moretdname, sizeof(name));
	if (name[0] == '\0' || strcmp(kipp->ki_comm, name) == 0) {
		name[0] = '-';
		name[1] = '\0';
	}

	return (name);
}

<<<<<<< HEAD
static const struct procstat_cmd *
getcmd(const char *str)
{
	const struct procstat_cmd *cmd;
	size_t i, l;
	int cmp, s;

	if (str == NULL)
		return (NULL);
	cmd = NULL;
	if ((l = strlen(str)) == 0)
		return (getcmd("basic"));
	s = l > 1 && strcasecmp(str + l - 1, "s") == 0;
	for (i = 0; i < nitems(cmd_table); i++) {
		/*
		 * After the first match substring matches are disabled,
		 * allowing subsequent full matches to take precedence.
		 */
		if (cmd == NULL && (cmd_table[i].cmp & PS_CMP_SUBSTR))
			cmp = strncasecmp(str, cmd_table[i].command, l -
			    ((cmd_table[i].cmp & PS_CMP_PLURAL) && s ? 1 : 0));
		else if ((cmd_table[i].cmp & PS_CMP_PLURAL) && s &&
		    l == strlen(cmd_table[i].command) + 1)
			cmp = strncasecmp(str, cmd_table[i].command, l - 1);
		else
			cmp = strcasecmp(str, cmd_table[i].command);
		if (cmp == 0)
			cmd = &cmd_table[i];
	}
	return (cmd);
}

=======
>>>>>>> 930409367ecf72a59ee5666730e1b84ac90527b2
int
main(int argc, char *argv[])
{
	int ch, interval;
	int i;
	struct kinfo_proc *p;
	const struct procstat_cmd *cmd;
	struct procstat *prstat, *cprstat;
	long l;
	pid_t pid;
	char *dummy;
	char *nlistf, *memf;
	int aflag;
	int cnt;

	interval = 0;
	cmd = NULL;
	memf = nlistf = NULL;
	aflag = 0;
	argc = xo_parse_args(argc, argv);

<<<<<<< HEAD
	while ((ch = getopt(argc, argv, "abCcefHhijkLlM:N:nrSstvw:x")) != -1) {
=======
	while ((ch = getopt(argc, argv, "abCcefHhijklLM:N:nrSstvw:x")) != -1) {
>>>>>>> 930409367ecf72a59ee5666730e1b84ac90527b2
		switch (ch) {
		case 'a':
			aflag++;
			break;
		case 'b':
			if (cmd != NULL)
				usage();
			cmd = getcmd("binary");
			break;
		case 'C':
			procstat_opts |= PS_OPT_CAPABILITIES;
			break;
		case 'c':
			if (cmd != NULL)
				usage();
			cmd = getcmd("arguments");
			break;
		case 'e':
			if (cmd != NULL)
				usage();
			cmd = getcmd("environment");
			break;
		case 'f':
			if (cmd != NULL)
				usage();
			cmd = getcmd("files");
			break;
		case 'H':
			procstat_opts |= PS_OPT_PERTHREAD;
			break;
		case 'h':
			procstat_opts |= PS_OPT_NOHEADER;
			break;
		case 'i':
			if (cmd != NULL)
				usage();
			cmd = getcmd("signals");
			break;
		case 'j':
			if (cmd != NULL)
				usage();
			cmd = getcmd("tsignals");
			break;
		case 'k':
			if (cmd != NULL && cmd->cmd == procstat_kstack) {
				if ((procstat_opts & PS_OPT_VERBOSE) != 0)
					usage();
				procstat_opts |= PS_OPT_VERBOSE;
			} else {
				if (cmd != NULL)
					usage();
				cmd = getcmd("kstack");
			}
			break;
		case 'L':
			if (cmd != NULL)
				usage();
			cmd = getcmd("ptlwpinfo");
			break;
		case 'l':
			if (cmd != NULL)
				usage();
			cmd = getcmd("rlimit");
			break;
<<<<<<< HEAD
		case 'M':
			memf = optarg;
=======

		case 'L':
			Lflag++;
			xocontainer = "ptlwpinfo";
			break;

		case 'n':
			nflag++;
>>>>>>> 930409367ecf72a59ee5666730e1b84ac90527b2
			break;
		case 'N':
			nlistf = optarg;
			break;
		case 'n':
			procstat_opts |= PS_OPT_SIGNUM;
			break;
		case 'r':
			if (cmd != NULL)
				usage();
			cmd = getcmd("rusage");
			break;
		case 'S':
			if (cmd != NULL)
				usage();
			cmd = getcmd("cpuset");
			break;
		case 's':
			if (cmd != NULL)
				usage();
			cmd = getcmd("credentials");
			break;
		case 't':
			if (cmd != NULL)
				usage();
			cmd = getcmd("threads");
			break;
		case 'v':
			if (cmd != NULL)
				usage();
			cmd = getcmd("vm");
			break;
		case 'w':
			l = strtol(optarg, &dummy, 10);
			if (*dummy != '\0')
				usage();
			if (l < 1 || l > INT_MAX)
				usage();
			interval = l;
			break;
		case 'x':
			if (cmd != NULL)
				usage();
			cmd = getcmd("auxv");
			break;
		case '?':
		default:
			usage();
		}

	}
	argc -= optind;
	argv += optind;

	if (cmd == NULL && argv[0] != NULL && (cmd = getcmd(argv[0])) != NULL) {
		if ((procstat_opts & PS_SUBCOMMAND_OPTS) != 0)
			usage();
		if (cmd->opt != NULL) {
			optreset = 1;
			optind = 1;
			cmd->opt(argc, argv);
			argc -= optind;
			argv += optind;
		} else {
			argc -= 1;
			argv += 1;
		}
	} else {
		if (cmd == NULL)
			cmd = getcmd("basic");
		if (cmd->cmd != procstat_files &&
		    (procstat_opts & PS_OPT_CAPABILITIES) != 0)
			usage();
	}

	/* Must specify either the -a flag or a list of pids. */
	if (!(aflag == 1 && argc == 0) && !(aflag == 0 && argc > 0))
		usage();

	if (memf != NULL)
		prstat = procstat_open_kvm(nlistf, memf);
	else
		prstat = procstat_open_sysctl();
	if (prstat == NULL)
		xo_errx(1, "procstat_open()");
	do {
		xo_set_version(PROCSTAT_XO_VERSION);
		xo_open_container("procstat");
		xo_open_container(cmd->xocontainer);

		if (aflag) {
			p = procstat_getprocs(prstat, KERN_PROC_PROC, 0, &cnt);
			if (p == NULL)
				xo_errx(1, "procstat_getprocs()");
			kinfo_proc_sort(p, cnt);
			for (i = 0; i < cnt; i++) {
				procstat(cmd, prstat, &p[i]);

				/* Suppress header after first process. */
				procstat_opts |= PS_OPT_NOHEADER;
				xo_flush();
			}
			procstat_freeprocs(prstat, p);
		}
		for (i = 0; i < argc; i++) {
			l = strtol(argv[i], &dummy, 10);
			if (*dummy == '\0') {
				if (l < 0)
					usage();
				pid = l;

				p = procstat_getprocs(prstat, KERN_PROC_PID,
				    pid, &cnt);
				if (p == NULL)
					xo_errx(1, "procstat_getprocs()");
				if (cnt != 0)
					procstat(cmd, prstat, p);
				procstat_freeprocs(prstat, p);
			} else {
				cprstat = procstat_open_core(argv[i]);
				if (cprstat == NULL) {
					warnx("procstat_open()");
					continue;
				}
				p = procstat_getprocs(cprstat, KERN_PROC_PID,
				    -1, &cnt);
				if (p == NULL)
					xo_errx(1, "procstat_getprocs()");
				if (cnt != 0)
					procstat(cmd, cprstat, p);
				procstat_freeprocs(cprstat, p);
				procstat_close(cprstat);
			}
			/* Suppress header after first process. */
			procstat_opts |= PS_OPT_NOHEADER;
		}

		xo_close_container(cmd->xocontainer);
		xo_close_container("procstat");
		xo_finish();
		if (interval)
			sleep(interval);
	} while (interval);

	procstat_close(prstat);

	exit(0);
}

void
cmdopt_none(int argc, char * const argv[])
{
	int ch;

	while ((ch = getopt(argc, argv, "")) != -1) {
		switch (ch) {
		case '?':
		default:
			usage();
		}
	}
}

void
cmdopt_verbose(int argc, char * const argv[])
{
	int ch;

	while ((ch = getopt(argc, argv, "v")) != -1) {
		switch (ch) {
		case 'v':
			procstat_opts |= PS_OPT_VERBOSE;
			break;
		case '?':
		default:
			usage();
		}
	}
}

void
cmdopt_signals(int argc, char * const argv[])
{
	int ch;

	while ((ch = getopt(argc, argv, "n")) != -1) {
		switch (ch) {
		case 'n':
			procstat_opts |= PS_OPT_SIGNUM;
			break;
		case '?':
		default:
			usage();
		}
	}
}

void
cmdopt_rusage(int argc, char * const argv[])
{
	int ch;

	while ((ch = getopt(argc, argv, "Ht")) != -1) {
		switch (ch) {
		case 'H':
			/* FALLTHROUGH */
		case 't':
			procstat_opts |= PS_OPT_PERTHREAD;
			break;
		case '?':
		default:
			usage();
		}
	}
}

void
cmdopt_files(int argc, char * const argv[])
{
	int ch;

	while ((ch = getopt(argc, argv, "C")) != -1) {
		switch (ch) {
		case 'C':
			procstat_opts |= PS_OPT_CAPABILITIES;
			break;
		case '?':
		default:
			usage();
		}
	}
}

void
cmdopt_cpuset(int argc, char * const argv[])
{

	procstat_opts |= PS_OPT_PERTHREAD;
	cmdopt_none(argc, argv);
}
